mobile/examples/phi-3/android/app/src/main/java/ai/onnxruntime/genai/demo/ModelDownloader.java

```diff
@@ -14,52 +17,88 @@
+    void onProgress(long lastBytesRead, long bytesRead, long bytesTotal);
     void onDownloadComplete() throws GenAIException;
   }
 
-  public static void downloadModel(Context context, String url, String fileName, DownloadCallback callback) {
+  public static void downloadModel(Context context, List<Pair<String, String>> urlFilePairs, DownloadCallback callback) {
     try {
-      File file = new File(context.getFilesDir(), fileName);
-      File tempFile = new File(context.getFilesDir(), fileName + ".tmp");
-      URL modelUrl = new URL(url);
-      HttpURLConnection connection = (HttpURLConnection) modelUrl.openConnection();
-      connection.connect();
-
-      // Check if response code is OK
-      if (connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
-        InputStream inputStream = connection.getInputStream();
-        FileOutputStream outputStream = new FileOutputStream(tempFile);
-
-        byte[] buffer = new byte[4096];
-        int bytesRead;
-        while ((bytesRead = inputStream.read(buffer)) != -1) {
-          outputStream.write(buffer, 0, bytesRead);
-        }
 
-        outputStream.flush();
-        outputStream.close();
-        inputStream.close();
+      List<HttpURLConnection> connections = new ArrayList<>();
+      long totalDownloadBytes = 0;
+      for (int i = 0; i < urlFilePairs.size(); i++) {
+        String url = urlFilePairs.get(i).first;
+        URL modelUrl = new URL(url);
+        HttpURLConnection connection = (HttpURLConnection) modelUrl.openConnection();
+        connections.add(connection);
+        long totalFileSize = connection.getHeaderFieldLong("Content-Length",-1);
+        totalDownloadBytes += totalFileSize;
+      }
+
+      long totalBytesRead = 0;
+      for (int i = 0; i < urlFilePairs.size(); i++) {
+        String fileName = urlFilePairs.get(i).second;
+        HttpURLConnection connection = connections.get(i);
+
+        File file = new File(context.getFilesDir(), fileName);
+        File tempFile = new File(context.getFilesDir(), fileName + ".tmp");
+        Log.d(TAG, "Downloading file: " + fileName);
+        connection.connect();
+
+        // Check if response code is OK
+        if (connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
+          InputStream inputStream = connection.getInputStream();
+          FileOutputStream outputStream = new FileOutputStream(tempFile);
 
-        // File downloaded successfully
-        if (tempFile.renameTo(file)) {
-          Log.d(TAG, "File downloaded successfully");
-          if (callback != null) {
-            callback.onDownloadComplete();
+          long begin = System.currentTimeMillis();
+
+          byte[] buffer = new byte[4096];
+          int bytesRead;
+
+          while ((bytesRead = inputStream.read(buffer)) != -1) {
+            outputStream.write(buffer, 0, bytesRead);
+            if (callback != null) {
+              callback.onProgress(totalBytesRead, totalBytesRead + bytesRead, totalDownloadBytes);
```

Comment:
I think we can assume that `bytesRead` will always be greater than zero within this loop. i.e., `onProgress` will be called only if there is actually progress.
https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A-

Reply:
top comment: model != null should be an atomic comparison, so I think it should be safe in this context. The only dangerous thing is if it becomes null after being non null, but those should be all processed in the UI thread.

bottom comment: I think that's the correct assumption. It becomes -1 when we reach the end.